data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.source.rds.stream;
+
+import com.github.shyiko.mysql.binlog.BinaryLogClient;
+import com.github.shyiko.mysql.binlog.event.DeleteRowsEventData;
+import com.github.shyiko.mysql.binlog.event.EventType;
+import com.github.shyiko.mysql.binlog.event.TableMapEventData;
+import com.github.shyiko.mysql.binlog.event.TableMapEventMetadata;
+import com.github.shyiko.mysql.binlog.event.UpdateRowsEventData;
+import com.github.shyiko.mysql.binlog.event.WriteRowsEventData;
+import org.opensearch.dataprepper.buffer.common.BufferAccumulator;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.opensearch.OpenSearchBulkActions;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.source.rds.RdsSourceConfig;
+import org.opensearch.dataprepper.plugins.source.rds.converter.StreamRecordConverter;
+import org.opensearch.dataprepper.plugins.source.rds.model.TableMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class BinlogEventListener implements BinaryLogClient.EventListener {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BinlogEventListener.class);
+
+    static final Duration BUFFER_TIMEOUT = Duration.ofSeconds(60);
+    static final int DEFAULT_BUFFER_BATCH_SIZE = 1_000;
+
+    /**
+     * TableId to TableMetadata mapping
+     */
+    private final Map<Long, TableMetadata> tableMetadataMap;
+
+    private final StreamRecordConverter recordConverter;
+    private final BufferAccumulator<Record<Event>> bufferAccumulator;
+    private final List<String> tableNames;
+    private final String s3Prefix;
+
+    public BinlogEventListener(final Buffer<Record<Event>> buffer, final RdsSourceConfig sourceConfig) {
```

Comment:
How does this EventListener work in regards to checkpointing and reprocessing in case of a failure with acknowledgments?

Reply:
Good question. I plan to add that in followup PRs. 

The basic idea is that we will do checkpointing with latest acknowledged binlog coordinate (binlog filename + position, similar to resume token in docdb or sequence number in ddb). To restart from a checkpoint, we can reconnect binlog client from the binlog coordinate in the checkpoint and we will start getting events since the checkpoint, EventListener will then continue to process events.